db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/TriggerEventActivator.java

```diff
@@ -31,49 +31,69 @@
 
 package com.splicemachine.db.impl.sql.execute;
 
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-import java.util.Vector;
-
 import com.splicemachine.db.catalog.UUID;
 import com.splicemachine.db.iapi.error.StandardException;
+import com.splicemachine.db.iapi.jdbc.ConnectionContext;
+import com.splicemachine.db.iapi.services.context.ContextService;
 import com.splicemachine.db.iapi.services.io.FormatableBitSet;
 import com.splicemachine.db.iapi.sql.Activation;
+import com.splicemachine.db.iapi.sql.conn.ConnectionUtil;
 import com.splicemachine.db.iapi.sql.conn.LanguageConnectionContext;
 import com.splicemachine.db.iapi.sql.conn.StatementContext;
 import com.splicemachine.db.iapi.sql.dictionary.TriggerDescriptor;
 import com.splicemachine.db.iapi.sql.execute.CursorResultSet;
 
+import java.sql.SQLException;
+import java.util.*;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.splicemachine.db.impl.sql.execute.TriggerExecutionContext.pushLanguageConnectionContextToCM;
+
 /**
  * Responsible for firing a trigger or set of triggers based on an event.
  */
 public class TriggerEventActivator {
 
-    private LanguageConnectionContext lcc;
     private TriggerInfo triggerInfo;
     private TriggerExecutionContext tec;
     private Map<TriggerEvent, List<GenericTriggerExecutor>> statementExecutorsMap = new HashMap<>();
-    private Map<TriggerEvent, List<GenericTriggerExecutor>> rowExecutorsMap = new HashMap<>();
+    private Map<TriggerEvent, List<TriggerDescriptor>> rowExecutorsMap = new HashMap<>();
+    private Map<TriggerEvent, List<TriggerDescriptor>> rowConcurrentExecutorsMap = new HashMap<>();
     private Activation activation;
+    private ConnectionContext cc;
     private String statementText;
     private UUID tableId;
     private String tableName;
     private boolean tecPushed;
+    private boolean esvPushed;
+    private LanguageConnectionContext esvLCC1;
```

Comment:
Added this comment:
    // getLcc() may return a different LanguageConnectionContext at the time
    // we pop the triggerExecutionContext and executionStmtValidator, versus
    // at the time they were pushed.  Saving the original "LCC" at the time of
    // the push ensures that we pop from the correct LCC.  We sometimes need to 
    // save two versions LCCs because we may push to the LCC in the activation, 
    // and also to the LCC stored in the current ContextManager, if they happen 
    // to be different.  Trigger execution requires that both the activation and
    // the current ContextManager both have a triggerExecutionContext.